p2p/base_reactor.go

```diff
@@ -41,25 +46,151 @@ type Reactor interface {
+	// queue to avoid blocking. Incoming messages are then passed to a
+	// processing function. The default processing function unmarshals the
+	// messages in the order the sender sent them and then calls Receive on the
+	// reactor. The queue size and the processing function can be changed by
+	// passing options to the base reactor.
+	QueueUnprocessedEnvelope(e UnprocessedEnvelope)
 }
 
 // --------------------------------------
 
 type BaseReactor struct {
 	service.BaseService // Provides Start, Stop, .Quit
 	Switch              *Switch
+
+	incoming chan UnprocessedEnvelope
+
+	// processor is called with the incoming channel and is responsible for
+	// unmarshalling the messages and calling Receive on the reactor.
+	processor func(incoming <-chan UnprocessedEnvelope) error
 }
 
-func NewBaseReactor(name string, impl Reactor) *BaseReactor {
-	return &BaseReactor{
+type ReactorOptions func(*BaseReactor)
+
+func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReactor {
+	base := &BaseReactor{
 		BaseService: *service.NewBaseService(nil, name, impl),
 		Switch:      nil,
+		incoming:    make(chan UnprocessedEnvelope, 100),
+		processor:   DefaultProcessor(impl),
+	}
+
+	for _, opt := range opts {
+		opt(base)
+	}
+
+	go func() {
+		base.processor(base.incoming)
+	}()
+
+	return base
+}
+
+// WithProcessor sets the processor function for the reactor. The processor
+// function is called with the incoming channel and is responsible for
+// unmarshalling the messages and calling Receive on the reactor.
+func WithProcessor(processor func(<-chan UnprocessedEnvelope) error) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.processor = processor
+	}
+}
+
+// WithIncomingQueueSize sets the size of the incoming message queue for a
+// reactor.
+func WithIncomingQueueSize(size int) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.incoming = make(chan UnprocessedEnvelope, size)
 	}
 }
 
 func (br *BaseReactor) SetSwitch(sw *Switch) {
 	br.Switch = sw
 }
+
+// QueueUnprocessedEnvelope is called by the switch when an unprocessed
+// envelope is received. Unprocessed envelopes are immediately buffered in a
+// queue to avoid blocking. The size of the queue can be changed by passing
+// options to the base reactor.
+func (br *BaseReactor) QueueUnprocessedEnvelope(e UnprocessedEnvelope) {
+	br.incoming <- e
+}
+
+// DefaultProcessor unmarshals the message and calls Receive on the reactor.
+// This preservers the sender's original order for all messages.
+func DefaultProcessor(impl Reactor) func(<-chan UnprocessedEnvelope) error {
+	return func(incoming <-chan UnprocessedEnvelope) error {
+		implChannels := impl.GetChannels()
+
+		chIDs := make(map[byte]proto.Message, len(implChannels))
+		for _, chDesc := range implChannels {
+			chIDs[chDesc.ID] = chDesc.MessageType
+		}
+
+		var (
+			err error
+			msg proto.Message
+		)
+
+		for ue := range incoming {
```

Comment:
had some time to debug this tonight, and we're seeing a really weird issue. The messages are that are queued in this channel are unique, but they come out here, they aren't always :fearful: 

interesting... will continue to debug tmrw